Linux Installer

karottenreibe opened the pull request:
Addresses issue [TS-31573](https://cqse.atlassian.net/browse/TS-31573)

- [x] Changes are tested adequately
- [ ] Agent's README.md updated in case of user-visible changes
- [x] CHANGELOG.md updated
- [ ] Present new features in [N&N](https://wiki.cqse.eu/pages/viewpage.action?pageId=689566)
- [ ] [TGA Tutorial](https://docs.teamscale.com/tutorial/setting-up-tga-java) updated
- [ ] [TIA Tutorial](https://docs.teamscale.com/tutorial/tia-java/) updated

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.

karottenreibe commented:
Short update after a first manual test on a VM:

- some minor edge cases left to fix
- Tomcat sets `ProtectSystem=strict` in its systemd unit file, which makes /opt no longer writable (for log files and coverage files) https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html
- need to handle multiple processes trying to write to the same log file

karottenreibe commented:
Last manual test was a success, but I had to resolve some pretty big merge conflicts, so redoing the manual test to be sure

DreierF commented on installer/src/main/java/com/teamscale/profiler/installer/TeamscaleUtils.java:
```diff
@@ -0,0 +1,104 @@
+			throw new FatalInstallerError("Failed to connect via HTTPS to " + credentials.url
+					+ "\nPlease ensure that your Teamscale instance is reachable under " + credentials.url
+					+ " and that it is configured for HTTPS, not HTTP. E.g. open that URL in your"
+					+ " browser and verify that you can connect successfully.",
```
Might also happen if the server uses an internally signed SSL certificate that is not installed on the current machine, right?

DreierF commented on installer/src/test/java/com/teamscale/profiler/installer/MockTeamscale.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Mocks a redirect server: redirects all requests to another server.
+ */
+	public class MockTeamscale {
```
Formatting off

DreierF commented on system-tests/debug-logging-default-test/build.gradle.kts:
```diff
@@ -1,7 +1,15 @@
+		// the test checks if a log file is written, so we need to remove all leftover agent temp directories
+		// to avoid false-positive test failures
+		val leftoverAgentTempDirectories = file(System.getProperty("java.io.tmpdir")).listFiles()
+			.filter { it.name.contains("teamscale-java-profiler") }
```
The system tests are running in parallel. Doesn't this become a problem if another system test writes stuff to there?

karottenreibe replied:
ah, good point. Didn't think about this. I now changed the test to not use `/tmp` but a private directory as the system temp dir

DreierF commented on system-tests/teamscale-properties-test/build.gradle.kts:
```diff
@@ -0,0 +1,34 @@
+
+tasks.test {
+	/** These ports must match what is configured in the SystemTest class. */
+	val agentPort = 63301
```
This agent and Teamscale port is already used by HttpRedirectSystemTest. Please pick a different one to not cause issues when executing the tests in parallel

karottenreibe replied:
good point, I wasn't aware that we need to avoid conflicts here. I augmented the system test convention so ports are now picked automatically so they don't conflict. I hope I got this right ;-)

karottenreibe commented on agent/build.gradle.kts:
```diff
@@ -12,13 +12,15 @@ plugins {
 	com.teamscale.`logger-patch`
 }
 
+evaluationDependsOn(":installer")
```
I didn't think I'd see the day when I teach you something about Gradle :P :partying_face: ;-) (though I must admit I just googled this one)

karottenreibe commented on installer/src/main/java/com/teamscale/profiler/installer/InstallCommand.java:
```diff
@@ -0,0 +1,37 @@
+	@CommandLine.Parameters(index = "0", converter = HttpUrlTypeConverter.class, description = "The URL of your Teamscale instance.")
+	private HttpUrl teamscaleUrl;
+
+	@CommandLine.Parameters(index = "1", description = "The user used to access your Teamscale instance and upload coverage to your projects.")
```
tbh during the installation it can be any valid user. We can only check if the user can access Teamscale. Since during installation we don't have any particular project yet that we know should be accessed, we can't check permissions at this point other than "can authenticate with Teamscale".

karottenreibe commented on installer/src/main/java/com/teamscale/profiler/installer/Installer.java:
```diff
@@ -0,0 +1,192 @@
+	private static final String RERUN_ADVICE = getRerunAdvice();
+
+	private static String getRerunAdvice() {
+		if (SystemUtils.isWindows()) {
```
ah yes, I got carried away here and started thinking about how/if this could work on windows :P do you want me to remove it or leave it in for later when we implement Windows support? It's sort of dead code ATM, so can't be triggered (since we don't compile for Windows)

DreierF replied:
You can leave it. Was just confused :D

karottenreibe commented on agent/src/main/java/com/teamscale/jacoco/agent/util/AgentUtils.java:
```diff
@@ -2,14 +2,55 @@
+			return pid.toString();
+		} catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException |
+				 InvocationTargetException e) {
+			return ManagementFactory.getRuntimeMXBean().getName();
```
it might. I'm sanitizing this now

karottenreibe commented on agent/src/test/java/com/teamscale/jacoco/agent/options/AgentOptionsParserTest.java:
```diff
@@ -3,16 +3,40 @@
+		Path config = tempDir.resolve("config.properties");
+		Files.write(config, "debug=true".getBytes(StandardCharsets.UTF_8));
+		AgentOptionsParser parser = new AgentOptionsParser(new CommandLineLogger(), config.toString(), null);
+		AgentOptions options = parser.parse("debug=false");
```
we could, but this would be a bigger change to how the option parsing works since that doesnt keep track of how often an option occurs yet. So I'll create a ticket for that instead. I'll document that env options override command line options to at least make it clear in the docs

karottenreibe replied:
Created https://cqse.atlassian.net/browse/TS-36384 and moved documentation task to https://cqse.atlassian.net/browse/TS-36383

karottenreibe commented on agent/src/main/java/com/teamscale/jacoco/agent/PreMain.java:
```diff
@@ -31,11 +34,32 @@ public class PreMain {
+	private static final String LOCKING_SYSTEM_PROPERTY = "TEAMSCALE_JAVA_PROFILER_ATTACHED";
+
+	/** Environment variable from which to read the config file to use. */
+	private static final String CONFIG_ENVIRONMENT_VARIABLE = "TEAMSCALE_JAVA_PROFILER_CONFIG";
```
Moved to https://cqse.atlassian.net/browse/TS-36383

karottenreibe commented on installer/build.gradle.kts:
```diff
@@ -0,0 +1,54 @@
+
+dependencies {
+	// we need this older version since newer versions are Kotlin-implemented and don't play nice with GraalVM
+	implementation("com.squareup.okhttp3:okhttp:3.14.2")
```
No, official support will come with okhttp 5 which is currently alpha. Added a link to the changelog instead :-)

karottenreibe commented on installer/src/main/java/com/teamscale/profiler/installer/Installer.java:
```diff
@@ -0,0 +1,192 @@
+			if (SystemUtils.isLinux()) {
+				System.out.println("To use the profiler in your current session, please log out and back in.");
+			}
+			System.out.println("To activate the profiler for an application, set the environment variable:"
+					+ "\nTEAMSCALE_JAVA_PROFILER_CONFIG"
+					+ "\nIts value must be the path to a valid profiler configuration file."
+					+ "\nThen, restart your application (for web applications: restart the app server).");
```
Moved to https://cqse.atlassian.net/browse/TS-36383

DreierF commented on buildSrc/src/main/kotlin/SystemTestUtils.kt:
```diff
@@ -0,0 +1,31 @@
+}
+
+/** The port that the fake Teamscale should use during the system test. Guaranteed conflict-free. */
+val Test.teamscalePort: Int by lazy { SystemTestUtils.pickFreePort() }
```
Very nice idea!
Unfortunately this won't work reliably, because Gradle already uses classloader isolation for subprojects that do not apply the exact same set of plugins, which is already the case for some system test modules. This will also be used more in the future with configuration caching and project isolation being enabled. The only reliable solution for this problem is to make `SystemTestUtils` a [shared build service](https://docs.gradle.org/current/userguide/build_services.html) (see also the related discussion on [Github](https://github.com/gradle/gradle/issues/14697))

karottenreibe replied:
I did my best but I'm not 100% sure I implemented this build service right. The docs are all about using the service in custom tasks but we're modifying the existing Test task, so that doesn't seem to apply. At least in my local setup, it picks all distinct ports, but I don't think it ran anything in parallel anyways, so better review this extra carefully ;-)

DreierF commented on installer/build.gradle.kts:
```diff
@@ -18,6 +18,7 @@ application {
 dependencies {
 	// we need this older version since newer versions are Kotlin-implemented and don't play nice with GraalVM
+	// okhttp 5 will add official GraalVM support but is currently still alpha https://square.github.io/okhttp/changelogs/changelog/
 	implementation("com.squareup.okhttp3:okhttp:3.14.2")
```
You can also add this to the `libs.versions.toml` and add add the following to the `renovate.json`:
```
    {
      "matchPackageNames": [  "com.squareup.okhttp3:okhttp" ],
      "allowedVersions": ">= 5"
    },
```
Then we'll get a PR once 5.0 is available and we cannot miss it :)

karottenreibe replied:
hm, not sure this is so easy since the agent already uses a 4.* version of okhttp. So this would mean downgrading the agent's library, right? Which probably means changing a bunch of code since 3 and 4 are not API compatible, 4 being a complete rewrite in Kotlin.

DreierF replied:
Ah sorry, I missed that we already use v4.

DreierF commented on buildSrc/src/main/kotlin/SystemTestPorts.kt:
```diff
@@ -0,0 +1,69 @@
+val Test.teamscalePort: Int
+	get() {
+		synchronized(teamscalePortMap) {
+			return teamscalePortMap.getOrPut(this) { SystemTestPorts.pickFreePort() }
```
This might work, but I'm not really sure to be honest as we are now sharing one instance of the Provider potentially across projects. I would suggest the following:

Store the chosen provider and the chosen port as an extension property directly on the test instance so that we don't need to keep the test instances in a map, which kind of feels dangerous.
```
var Test.portProvider: Provider<SystemTestPorts>
	get() = extra["portProvider"] as Provider<SystemTestPorts>
	set(value) {
		extra["portProvider"] = value
	}

var Test.teamscalePort: Int
	get() = extra["teamscalePort"] as Int
	set(value) {
		extra["teamscalePort"] = value
	}
```

In `com.teamscale.system-test-convention.gradle.kts` you can do:
```
val provider = SystemTestPorts.registerWith(project)

tasks.test {
//...
	usesService(provider)
	portProvider = provider
	teamscalePort = provider.get().pickFreePort()
	agentPort = provider.get().pickFreePort()
	systemProperties("agentPort" to agentPort, "teamscalePort" to teamscalePort)
}
```
and hence don't have to rely on the global provider instance anymore.

For the redirect/artifactory ports it becomes `val artifactoryPort = portProvider.get().pickFreePort()`

I committed the changes to your branch as I already had them locally anyway. Please have look if you agree: https://github.com/cqse/teamscale-jacoco-agent/pull/287/commits/fa72d4daa0131ec810d24f8e98eabc918587170d

karottenreibe replied:
looks excellent, thank you!

DreierF reviewed (changes requested):
Thanks for the rework! I pushed two commits one to disable the InstallerTests for non linux systems and the other one to refactor the Gradle code (see separate comment).

However there is still a problem with the test setup. When I run `./gradlew build` locally I get 
```
> Task :system-tests:artifactory-git-properties-detection:test FAILED
Failed to parse agent options: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
        at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
Caused by: com.teamscale.jacoco.agent.options.AgentOptionParseException: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
        at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:100)
        at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:67)
FATAL ERROR in native method: processing of -javaagent failed
        at com.teamscale.jacoco.agent.PreMain.getAndApplyAgentOptions(PreMain.java:92)
        at com.teamscale.jacoco.agent.PreMain.premain(PreMain.java:62)
```

This is because the test picks up the `teamscale.properties` file from a previous test run.
I guess this validation error could also become a problem if you would want to configure the agent to upload to artifactory in general.

karottenreibe commented:
> Thanks for the rework! I pushed two commits one to disable the InstallerTests for non linux systems and the other one to refactor the Gradle code (see separate comment).
> 
> However there is still a problem with the test setup. When I run `./gradlew build` locally I get
> 
> ```
> > Task :system-tests:artifactory-git-properties-detection:test FAILED
> Failed to parse agent options: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
> Exception in thread "main" java.lang.reflect.InvocationTargetException
>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>         at java.lang.reflect.Method.invoke(Method.java:498)
>         at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
>         at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
> Caused by: com.teamscale.jacoco.agent.options.AgentOptionParseException: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
>         at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:100)
>         at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:67)
> FATAL ERROR in native method: processing of -javaagent failed
>         at com.teamscale.jacoco.agent.PreMain.getAndApplyAgentOptions(PreMain.java:92)
>         at com.teamscale.jacoco.agent.PreMain.premain(PreMain.java:62)
> ```
> 
> This is because the test picks up the `teamscale.properties` file from a previous test run. I guess this validation error could also become a problem if you would want to configure the agent to upload to artifactory in general.

good catch. I now isolated the system tests so each of them uses their own copy of the shadowed agent jar, avoiding this problem

DreierF commented:
> good catch. I now isolated the system tests so each of them uses their own copy of the shadowed agent jar, avoiding this problem

But that means you'd still not be able to install the agent via the installer and upload to artifactory, right?

karottenreibe commented:
Correct, I haven't gotten to that part yet 😋

Florian Dreier ***@***.***> schrieb am Fr., 17. Nov. 2023,
17:47:

> good catch. I now isolated the system tests so each of them uses their own
> copy of the shadowed agent jar, avoiding this problem
>
> But that means you'd still not be able to install the agent via the
> installer and upload to artifactory, right?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cqse/teamscale-jacoco-agent/pull/287#issuecomment-1816756034>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAJPNYMGAC3YVMBZYI6KADYE6ILFAVCNFSM6AAAAAA2RKQM3GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMJWG42TMMBTGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

karottenreibe commented:
I fixed this now, but that was a bigger operation. To make sure I don't accidentally break anything, I added a bunch of additional tests for the option parsing for valid and invalid combinations.
I also refactored some of the option parsing code to enable my changes and make it easier to understand.
I also simplified some of the NWDI parsing code since that was unnecessarily complicated and also had a bug that the additional tests revealed.

Sorry for the big code change but it wasn't really possible otherwise AFAICS.

karottenreibe commented:
There's one architecture finding for which we IMO need to adjust the architecture to allow the dependency. So leaving that as is.

DreierF reviewed (approved):
Thanks for the cleanup! Really appreciate it!
